pkg/termination/termination.go

```diff
@@ -0,0 +1,176 @@
+package termination
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	machinev1 "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubectl/pkg/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	azureTerminationEndpointURL = "http://169.254.169.254/metadata/scheduledevents?api-version=2019-08-01"
+)
+
+// Handler represents a handler that will run to check the termination
+// notice endpoint and delete Machine's if the instance termination notice is fulfilled.
+type Handler interface {
+	Run(stop <-chan struct{}) error
+}
+
+// NewHandler constructs a new Handler
+func NewHandler(logger logr.Logger, cfg *rest.Config, pollInterval time.Duration, namespace, nodeName string) (Handler, error) {
+	machinev1.AddToScheme(scheme.Scheme)
+	c, err := client.New(cfg, client.Options{})
+	if err != nil {
+		return nil, fmt.Errorf("error creating client: %v", err)
+	}
+
+	pollURL, err := url.Parse(azureTerminationEndpointURL)
+	if err != nil {
+		// This should never happen
+		panic(err)
+	}
+
+	logger = logger.WithValues("node", nodeName, "namespace", namespace)
+
+	return &handler{
+		client:       c,
+		pollURL:      pollURL,
+		pollInterval: pollInterval,
+		nodeName:     nodeName,
+		namespace:    namespace,
+		log:          logger,
+	}, nil
+}
+
+// handler implements the logic to check the termination endpoint and delete the
+// machine associated with the node
+type handler struct {
+	client       client.Client
+	pollURL      *url.URL
+	pollInterval time.Duration
+	nodeName     string
+	namespace    string
+	log          logr.Logger
+}
+
+// Run starts the handler and runs the termination logic
+func (h *handler) Run(stop <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	errs := make(chan error, 1)
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+		errs <- h.run(ctx)
+	}()
+
+	select {
+	case <-stop:
+		cancel()
+		// Wait for run to stop
+		wg.Wait()
+		return nil
+	case err := <-errs:
+		cancel()
+		return err
+	}
+}
+
+func (h *handler) run(ctx context.Context) error {
+	machine, err := h.getMachineForNode(ctx)
```

Comment:
if there happen to be a transient error getting the node, we just fail and terminate the container?

Reply:
DaemonSet will restart the pod then. I think should not be an issue because we try to get the machine only once and at the beginning of `run()`. @JoelSpeed wdyt?

Reply:
It could certainly be improved, but this is how it is in AWS at present (likewise it exits 0 everywhere which it probably shouldn't). In the aws case, because it exits 0, there's no crash loop backoff on restart, if we change the exit codes here we should probably build in some retry logic if we think that's reasonable, but that is going to bloat the otherwise very small and simple piece of software. I wonder if it's worth waiting until we unify the termination handlers all in the MAO repo?

Reply:
Now that we log fatal if this fails, this could go into crash loop back off. Should we implement some retry logic for `run` so that we keep trying on errors rather than exiting the container?

Reply:
sgtm. Shouldn't this be an infinite poll?

Reply:
Make the whole function an infinite poll? Yeah that would work, just need to log errors but not return them.
Do you mean something like this?
```
errs <- wait.PollImmediateUntil(0, func() (bool, error) {
  err := h.run(ctx)
  if err != nil {
    log.Errorf(...)
    return false, nil
  }
  return true, nil
}, ctx.Done())
```

Reply:
I was thinking on just the bit getting the machine, so we block on `getMachineForNode(ctx)`. wdyt?

Reply:
Sure, that works. I think the benefit of my suggestion is that it would also cover Deletion failures, don't know if we would want to be retrying those?

Reply:
we could actually move `getMachineForNode(ctx)` to the bottom and cover both get and delete.

Reply:
Yeah sure, we would lose the Machine label on the logging, but that's ok, I think we are dropping fetching the Machine anyway when I get round to switching this to conditions (waiting on enhancement review for that)